JSIL/SourceMapBuilder.cs

```diff
@@ -0,0 +1,277 @@
+            if (string.IsNullOrEmpty(fromPath)) throw new ArgumentNullException("fromPath");
+            if (string.IsNullOrEmpty(toPath)) throw new ArgumentNullException("toPath");
+
+            Uri fromUri = new Uri(Path.GetFullPath(fromPath));
+            Uri toUri = new Uri(Path.GetFullPath(toPath));
+
+            if (fromUri.Scheme != toUri.Scheme) { return toUri; } // path can't be made relative.
+
+            return fromUri.MakeRelativeUri(toUri);
+        }
+
+        private sealed class Mapping
+        {
+            public int GeneratedLine { get; private set; }
+            public int GeneratedColumn { get; private set; }
+
+            public int OriginalLine { get; private set; }
+            public int OriginalColumn { get; private set; }
+
+            public string Source { get; private set; }
+            public string Name { get; private set; }
+
+            public Mapping(int generatedLine, int generatedColumn, int originalLine, int originalColumn, string source, string name)
+            {
+                GeneratedLine = generatedLine;
+                GeneratedColumn = generatedColumn;
+                OriginalLine = originalLine;
+                OriginalColumn = originalColumn;
+                Source = source;
+                Name = name;
+            }
+
+            private bool Equals(Mapping other)
+            {
+                return GeneratedLine == other.GeneratedLine && GeneratedColumn == other.GeneratedColumn && OriginalLine == other.OriginalLine && OriginalColumn == other.OriginalColumn && string.Equals(Source, other.Source) && string.Equals(Name, other.Name);
+            }
+
+            public override bool Equals(object obj)
+            {
+                if (ReferenceEquals(null, obj)) return false;
+                if (ReferenceEquals(this, obj)) return true;
+                if (obj.GetType() != this.GetType()) return false;
+                return Equals((Mapping) obj);
+            }
+
+            public override int GetHashCode()
+            {
+                unchecked
+                {
+                    var hashCode = GeneratedLine.GetHashCode();
+                    hashCode = (hashCode*397) ^ GeneratedColumn.GetHashCode();
+                    hashCode = (hashCode*397) ^ OriginalLine.GetHashCode();
+                    hashCode = (hashCode*397) ^ OriginalColumn.GetHashCode();
+                    hashCode = (hashCode*397) ^ (Source != null ? Source.GetHashCode() : 0);
+                    hashCode = (hashCode*397) ^ (Name != null ? Name.GetHashCode() : 0);
+                    return hashCode;
+                }
+            }
+
+            public static bool operator ==(Mapping left, Mapping right)
+            {
+                return Equals(left, right);
+            }
+
+            public static bool operator !=(Mapping left, Mapping right)
+            {
+                return !Equals(left, right);
+            }
+        }
+
+        private static class Base64Vlq
```

Comment:
Yes, it's Mozilla code (https://github.com/mozilla/source-map/), to be license compliant we need add header from: https://github.com/mozilla/source-map/blob/master/lib/base64-vlq.js.